tuf/repository_lib.py

```diff
@@ -1284,41 +1304,22 @@ def generate_targets_metadata(targets_directory, target_files, version,
 
   # Ensure the user is aware of a non-existent 'target_directory', and convert
   # it to its abosolute path, if it exists.
-  targets_directory = _check_directory(targets_directory)
+  check_exists = not use_existing_fileinfo
+  targets_directory = _check_directory(targets_directory, check_exists)
 
-  # Generate the fileinfo of all the target files listed in 'target_files'.
-  for target, fileinfo in six.iteritems(target_files):
+  if use_existing_fileinfo:
+    for target, fileinfo in six.iteritems(target_files):
 
-    # The root-most folder of the targets directory should not be included in
-    # target paths listed in targets metadata.
-    # (e.g., 'targets/more_targets/somefile.txt' -> 'more_targets/somefile.txt')
-    relative_targetpath = target
+      # Ensure all fileinfo entries in target_files have a non-empty hashes dict
+      if not fileinfo.get('hashes', None):
```

Comment:
Important: This should also fail if 'length' is missing.

Reply:
Great catch, thank you. I've squashed a change to implement this into the patch.